source/common/config/watch_map.h

```diff
@@ -0,0 +1,121 @@
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in added_.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in removed_.
+  AddedRemoved updateWatchInterest(Watch* watch,
+                                   const std::set<std::string>& update_to_these_names);
+
+  // SubscriptionCallbacks
+  virtual void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+                              const std::string& version_info) override;
+  virtual void
+  onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                 const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                 const std::string& system_version_info) override;
+
+  virtual void onConfigUpdateFailed(const EnvoyException* e) override;
+
+  virtual std::string resourceName(const ProtobufWkt::Any&) override {
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  }
+
+private:
+  friend struct Watch;
+  // Expects that the watch to be removed has already had all of its resource names removed via
+  // updateWatchInterest().
+  void removeWatch(Watch* watch);
+
+  // Given a list of names that are new to an individual watch, returns those names that are in fact
+  // new to the entire subscription.
+  std::set<std::string> findAdditions(const std::vector<std::string>& newly_added_to_watch,
+                                      Watch* watch);
+
+  // Given a list of names that an individual watch no longer cares about, returns those names that
+  // in fact the entire subscription no longer cares about.
+  std::set<std::string> findRemovals(const std::vector<std::string>& newly_removed_from_watch,
+                                     Watch* watch);
+
+  // Returns the union of watch_interest_[resource_name] and wildcard_watches_.
+  absl::flat_hash_set<Watch*> watchesInterestedIn(const std::string& resource_name);
+
+  absl::flat_hash_set<Watch*> watches_;
+
+  // Watches whose interest set is currently empty, which is interpreted as "everything".
+  absl::flat_hash_set<Watch*> wildcard_watches_;
+
+  // Maps a resource name to the set of watches interested in that resource. Has two purposes:
+  // 1) Acts as a reference count; no watches care anymore ==> the resource can be removed.
+  // 2) Enables efficient lookup of all interested watches when a resource has been updated.
+  absl::flat_hash_map<std::string, absl::flat_hash_set<Watch*>> watch_interest_;
+
+  WatchMap(const WatchMap&) = delete;
+  WatchMap& operator=(const WatchMap&) = delete;
+};
+
+struct Watch {
+  Watch(WatchMap& owning_map, SubscriptionCallbacks& callbacks)
+      : owning_map_(owning_map), callbacks_(callbacks) {}
+  ~Watch() { owning_map_.removeWatch(this); }
```

Comment:
removeWatch() requires that the watch's resource interest has already been set to empty. Over in my [ADS branch](https://github.com/fredlas/envoy/tree/ADS_incremental) I currently have DeltaSubscriptionImpl's destuctor doing that set-to-empty, followed by a watch_.reset().

The Watch's destructor here could take care of setting its own resource interest to empty, but that would require the WatchMap to hold a reference to the GrpcMux that owns it. Does that sound worth it?

Reply:
Nevermind, I looked into this, and it would require the Watch or Map to know about type_url, which is definitely not a clean structure. Furthermore, I'm starting to think all of the resource watching RAIIness should be out in Subscription, in which case DeltaSubscriptionImpl's dtor doing this wouldn't actually be messy.